[analyzer] Refine invalidation caused by `fread`

steakhal opened the pull request:
This change enables more accurate modeling of the write effects of `fread`. In particular, instead of invalidating the whole buffer, in a best-effort basis, we would try to invalidate the actually accesses elements of the buffer. This preserves the previous value of the buffer of the unaffected slots. As a result, diagnose more uninitialized buffer uses for example.

Currently, this refined invalidation only triggers for `fread` if and only if the `count` parameter and the buffer pointer's index component are concrete or perfectly-constrained symbols.
Additionally, if the `fread` would read more than 64 elements, the whole buffer is invalidated as before. This is to have safeguards against performance issues.

Refer to the comments of the assertions in the following example to see the changes in the diagnostics:

```c++
void demo() {
  FILE *fp = fopen("/home/test", "rb+");
  if (!fp) return;
  int buffer[10]; // uninitialized
  int read_items = fread(buffer+1, sizeof(int), 5, fp);
  if (5 == read_items) {
    int v1 = buffer[1]; // Unknown value but not garbage.
    clang_analyzer_isTainted(v1); // expected-warning {{YES}} <-- Would be "NO" without this patch.
    clang_analyzer_dump(v1); // expected-warning {{conj_}} <-- Not a "derived" symbol, so it's directly invalidated now.
    int v0 = buffer[0]; // expected-warning {{Assigned value is garbage or undefined}} <-- Had no report here before.
    (void)(v1 + v0);
  } else {
    // If 'fread' had an error.
    int v0 = buffer[0]; // expected-warning {{Assigned value is garbage or undefined}} <-- Had no report here before.
    (void)v0;
  }
  fclose(fp);
}
```

CPP-3247, CPP-3802

Patch by Marco Borgeaud (marco-antognini-sonarsource)

steakhal committed:
[analyzer] Refine invalidation caused by `fread`

This change enables more accurate modeling of the write effects of `fread`.
In particular, instead of invalidating the whole buffer, in a best-effort
basis, we would try to invalidate the actually accesses elements of the buffer.
This preserves the previous value of the buffer of the unaffected slots.
As a result, diagnose more uninitialized buffer uses for example.

Currently, this refined invalidation only triggers for `fread` if and
only if the `count` parameter and the buffer pointer's index component
are concrete or perfectly-constrained symbols.
Additionally, if the `fread` would read more than 64 elements, the whole
buffer is invalidated as before. This is to have safeguards against
performance issues.

Refer to the comments of the assertions in the following example to see
the changes in the diagnostics:

```c++
void demo() {
  FILE *fp = fopen("/home/test", "rb+");
  if (!fp) return;
  int buffer[10]; // uninitialized
  int read_items = fread(buffer+1, sizeof(int), 5, fp);
  if (5 == read_items) {
    int v1 = buffer[1]; // Unknown value but not garbage.
    clang_analyzer_isTainted(v1); // expected-warning {{YES}} <-- Would be "NO" without this patch.
    clang_analyzer_dump(v1); // expected-warning {{conj_}} <-- Not a "derived" symbol, so it's directly invalidated now.
    int v0 = buffer[0]; // expected-warning {{Assigned value is garbage or undefined}} <-- Had no report here before.
    (void)(v1 + v0);
  } else {
    // If 'fread' had an error.
    int v0 = buffer[0]; // expected-warning {{Assigned value is garbage or undefined}} <-- Had no report here before.
    (void)v0;
  }
  fclose(fp);
}
```

[CPP-3247](https://sonarsource.atlassian.net/browse/CPP-3247)

Patch by Marco Borgeaud (marco-antognini-sonarsource)

NagyDonat reviewed (commented):
Thanks for publishing this commit, it's a nice refinement of the modelling capabilities, and mostly LGTM.

In the inline comments I added some very minor remarks and one question about the handling of a `fread` that reads into the beginning of an array region.

balazske commented on clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp:
```diff
@@ -717,18 +717,71 @@ const ExplodedNode *StreamChecker::getAcquisitionSite(const ExplodedNode *N,
+  EscapingVals.reserve(*CountVal);
+
+  RegionAndSymbolInvalidationTraits ITraits;
+  for (auto Idx : llvm::seq(*StartIndexVal, *StartIndexVal + *CountVal)) {
```
This loop does not work if the type of the array is not the same as the "size" parameter passed to `fread`.:
```
int buffer[100];
fread(buffer + 1, 3, 5, file);
```
In this case 3*5 bytes should be read by `fread` into the array. If `sizeof(int)==4` 4 elements should be invalidated in the buffer starting from index 1.

steakhal replied:
Yes, this is a problem.
To properly invalidate element-wise, I must bind the correctly typed conjured symbols so that subsequent lookups in the store would hit the entry.
Consequently, I can't consider the buffer as a char array and invalidate each chars in the range.
I must use `ElementRegions`, thus I need properly aligned item-wise iteration and invalidation.

So, I could try to convert the element size and item count into a start byte offset and length in bytes. After this I could check if the start bytes offset would land on an item boundary and if the length would cover a whole element object.

This way we could cover the case like:
```
int buffer[100]; // assuming 4 byte ints.
fread(buffer + 1, 2, 10, file); // Reads 20 bytes of data, which is dividable by 4, thus we can think of the 'buffer[1:6]' elements as individually invalidated.
``` 

WDYT?

balazske commented on clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp:
```diff
@@ -717,18 +717,71 @@ const ExplodedNode *StreamChecker::getAcquisitionSite(const ExplodedNode *N,
+/// This is basically a refinement of the more generic 'escapeArgs' or
+/// the plain old 'invalidateRegions'.
+/// This only works if the \p StartIndex and \p Count are concrete or
+/// perfectly-constrained.
```
This comment is not very useful. References to other functions are not such important here, and the information that `Buffer` is used as array with specific type and elements in it are invalidated is missing.

NagyDonat replied:
>  References to other functions are not such important here
Actually, I think it's useful to mention related functions.

balazske commented on clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp:
```diff
@@ -937,8 +990,21 @@ void StreamChecker::evalFreadFwrite(const FnDescription *Desc,
+      const MemRegion *Buffer = BufferFirstElem->getSuperRegion();
+      QualType ElemTy = BufferFirstElem->getElementType();
+      SVal FirstAccessedItem = BufferFirstElem->getIndex();
+      SVal ItemCount = Call.getArgSVal(2);
```
Here is some redundancy in getting the argument values, `SizeVal` and `NMembVal` are already available and could be used by the function (if it would take `NonLoc`).

steakhal committed:
NFC Prefer dyn_cast over isa&cast

https://github.com/llvm/llvm-project/pull/93408#discussion_r1615909522

steakhal committed:
NFC Spell out std::optional

https://github.com/llvm/llvm-project/pull/93408#discussion_r1615913100

steakhal committed:
NFC Fix test name typos

https://github.com/llvm/llvm-project/pull/93408#discussion_r1615919113
https://github.com/llvm/llvm-project/pull/93408#discussion_r1615919288

steakhal committed:
NFC Move the test file to C

https://github.com/llvm/llvm-project/pull/93408#discussion_r1615930654

steakhal commented on clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp:
```diff
@@ -937,8 +990,21 @@ void StreamChecker::evalFreadFwrite(const FnDescription *Desc,
-    State = escapeArgs(State, C, Call, {0});
+  if (IsFread && !E.isStreamEof()) {
+    // Try to invalidate the individual elements.
+    if (const auto *BufferFirstElem =
+            dyn_cast_or_null<ElementRegion>(Call.getArgSVal(0).getAsRegion())) {
```
In case `fread` reads to the beginning of a buffer, we won't have an `ElementRegion`, thus the heuristic for eagerly binding the invalidated elements won't trigger. This is unfortunate, but you can think of this as we keep the previous behavior.
To circumvent this, I'd need to know the type for for the pointee.
This would imply that I should special-case `TypedValueRegion` and `SymbolicRegion`.

I'll think about it.

steakhal committed:
NFC Reuse existing variable

steakhal commented:
Fixed most NFC typos and suggestions.
Let's continue the discussion.

steakhal committed:
Handle TypedValueRegions and SymbolicRegions at offset 0

steakhal committed:
Demonstrate weird fread handling

balazske commented on clang/test/Analysis/fread.c:
```diff
@@ -0,0 +1,412 @@
+}
+
+void test_partial_elements_read(void) {
+  clang_analyzer_dump(sizeof(int)); // expected-warning {{4 S32b}}
```
Maybe some (buildbot) platforms are different?

steakhal replied:
I pinned the target triple in the RUN line to account for this.

balazske commented on clang/test/Analysis/fread.c:
```diff
@@ -0,0 +1,412 @@
+    if (20 == fread(buffer + 1, 3, 20, fp)) {
+      clang_analyzer_dump(buffer[0]); // expected-warning{{1 S32b}}
+      clang_analyzer_dump(buffer[20]); // expected-warning{{conj_}}
+      clang_analyzer_dump(buffer[21]); // expected-warning{{1st function call argument is an uninitialized value}}
```
Are these index values correct? `buffer[15]` should be the last changed item, `buffer[16]` should be 3, and `buffer[17]` is the uninitialized value.

steakhal replied:
Nice catch. I made a bug when calling `escapeByStartIndexAndCount`. I should have passed `NumElementsRead` instead of `CountVal`. Thanks for noticing this!

balazske commented on clang/test/Analysis/fread.c:
```diff
@@ -0,0 +1,412 @@
+    long buffer[10];
+    // Cannot reason about index.
+    size_t res = fread(buffer + startIndex, sizeof(long), 5, fp);
+    if (5 == res) {
```
It would be better to explain why these checks are here in this way. Is it not enough to have only the longest `if` branch for all cases if all are the same?

balazske commented on clang/test/Analysis/fread.c:
```diff
@@ -0,0 +1,412 @@
+    if (4 == fread(asChar + 1, 1, 4, fp)) {
+      void clang_analyzer_printState(void);
+      clang_analyzer_printState();
+      clang_analyzer_dump(buffer[0]); // expected-warning{{3 S32b}} FIXME Reading a 'char' should not result in a 'S32b' value.
```
The FIXME is not understandable to me, if this is about the problem seen in the PR remark (problem with store and overlapping read) it would be better to make this comment more accurate.

steakhal replied:
I tried to highlight that reading a 8 bit wide `char` element should never result in a value of 32 bits.
And this is casued by the current Store implementation, as: it does a lookup and it returns whatever is associated with that read start bit offset. Given that the last write the Store saw was the 32 bit wide int 3, it will just return that - no matter if that bitwidth would make sense or not.

I've added the 8 bit assumption to the FIXME. I hope it's good now.

balazske commented on clang/test/Analysis/fread.c:
```diff
@@ -0,0 +1,412 @@
+  }
+}
+
+void compound_write2(void) {
```
It is not clear to me why some of the test functions are called "write" and some "read". It would be better to have all of them `read` in the name.

steakhal replied:
Probably the author thought that the `read()` function call will have a `write` effect to the buffer.
TBH I agree with you and I'll rename these.

balazske commented on clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp:
```diff
@@ -907,6 +945,73 @@ void StreamChecker::preWrite(const FnDescription *Desc, const CallEvent &Call,
+    bool DivisibleAccessSpan = (NumBytesRead % ElemSizeInChars) == 0;
+    int64_t NumElementsRead = NumBytesRead / ElemSizeInChars;
+    constexpr int MaxInvalidatedElementsLimit = 64;
+    if (DivisibleAccessSpan && NumElementsRead <= MaxInvalidatedElementsLimit) {
```
Is it possible to round the `NumElementsRead` upwards instead of skipping the escape, in the not divisible case?

steakhal replied:
Applied a patch to this code such that the last partial element is considered as a full access.
This is indeed better than doing a fallback and invalidating the full buffer.
Also modified the test to demonstrate this, but there is yet another bug in the current Store:
If we read from a byte offset of an element, we still get unknown as a result: `int buffer[2] = {2,3}` then reading from `((char*)buffer)[1]` results in unknown. [Compiler explorer](https://godbolt.org/z/dTKGnYo9W).

balazske reviewed (commented):
It looks relatively good, I found only smaller issues.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp:
```diff
@@ -907,6 +945,73 @@ void StreamChecker::preWrite(const FnDescription *Desc, const CallEvent &Call,
+  auto Zero = [&SVB] {
+    BasicValueFactory &BVF = SVB.getBasicValueFactory();
+    return nonloc::ConcreteInt(BVF.getIntValue(0, /*isUnsigned=*/false));
+  };
```
This way it's lazily evaluated.

steakhal committed:
Remove leftover helper dumps

steakhal committed:
Fix argument passing bug

steakhal committed:
Simplify the dynamic_random_access_write test case

steakhal committed:
Ellaborate the bitwidth of chars

steakhal committed:
NFC Rename test cases to use "read" instead of "write"

steakhal committed:
Don't give up when 'fread' only partially reads the last element

steakhal commented:
Addressed all the feedback.

balazske commented on clang/test/Analysis/fread.c:
```diff
@@ -0,0 +1,443 @@
+    // We have an 'int' binding at offset 0 of value 3.
+    // We read 4 bytes at byte offset: 1,2,3,4.
+    if (4 == fread(asChar + 1, 1, 4, fp)) {
+      clang_analyzer_dump(buffer[0]); // expected-warning{{3 S32b}} FIXME Reading a 8 bit 'char' should not result in a 'S32b' value.
```
At this place not a `char` is read.

steakhal replied:
You are right. I copypasted the reasoning from the `asChar[0]`. In fact, the reason is:  `FIXME The int binding should have been partially overwritten by the read call. This definitely should not be 3.`.

Fixed the description of the FIXME.

balazske commented on clang/test/Analysis/fread.c:
```diff
@@ -0,0 +1,443 @@
+      clang_analyzer_dump(buffer[1]); // expected-warning{{conj_}}
+      clang_analyzer_dump(buffer[2]); // expected-warning{{5 S32b}}
+
+      clang_analyzer_dump(asChar[0]); // expected-warning{{3 S32b}} FIXME Reading a 8 bit 'char' should not result in a 'S32b' value.
```
Does it matter that clang_analyzer_dump has argument type `int`?

steakhal replied:
It shouldn't as the lvalue to rvalue conversion (aka. the load) operation should have been done before the call.
In contrast to this, CSA appears to contradict this assumption, and the `clang_analyzer_dump_char()` would have dumped `3 S8b`, which is the correct solution for little-endian systems - like the one I pinned as the target

I fixed this by dumping these as chars, and refined the expectation messages to meet this.

balazske commented on clang/test/Analysis/fread.c:
```diff
@@ -0,0 +1,443 @@
+  buffer[4] = 5;
+  FILE *fp = fopen("/home/test", "rb+");
+  if (fp) {
+    // 3*5: 15 bytes read; which is not exactly 4 integers, thus we invalidate the whole buffer.
```
This case now can be extended with a check that `buffer[6]` does not change.

steakhal replied:
I can't dump that because one byte (the last one) of that 'int' object is left uninitialized by the fread and reading uninit values will kill the execution path, making the rest of the checks unreachable.

This is caused by the current implementation which pretends that we only read into the first 4 elements of the buffer.

balazske replied:
I meant to add a `buffer[5] = 6;` at the begin and use `clang_analyzer_dump(buffer[5]);` after the read. If a value is initialized before the read it should not change to uninitialized.

balazske replied:
This changes the check for uninitialized `c[4+13]`? But probably only indexes 4, 8, 12, 16 in `c` (start of the integer values) are marked as initialized, and c[4+9] would be uninitialized too.

steakhal replied:
Ah, I think I forgot about this comment. Sorry, let me fix it.

steakhal committed:
Remove stale debug dumps

steakhal committed:
Refine tests by dumping chars only

steakhal committed:
Merge remote-tracking branch 'llvm/main' into refine-fread-escape

steakhal commented:
Let's do another round.

steakhal committed:
Check one more additional element in the array

steakhal committed:
Merge remote-tracking branch 'llvm/main' into refine-fread-escape

balazske reviewed (approved):
It looks now OK at least to my knowledge.

balazske commented:
It should be possible in `StreamChecker` to perform the invalidations even if the `StreamState` is not found (stream was not opened). Another possible solution is that a generic invalidation support is added to `StdLibraryFunctionsChecker` that can be used for stream related and other functions.

steakhal commented:
Thanks for the green light!

> Another possible solution is that a generic invalidation support is added to `StdLibraryFunctionsChecker` that can be used for stream related and other functions.

Yes, that would be so nice. However, it's out of scope for me this time.

aeubanks commented on clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp:
```diff
@@ -907,6 +945,76 @@ void StreamChecker::preWrite(const FnDescription *Desc, const CallEvent &Call,
+
+  if (const auto *ER = dyn_cast<ElementRegion>(R))
+    return ER->getIndex();
+  if (const auto *TR = dyn_cast<TypedValueRegion>(R))
```
unused variable (and ditto below), please use `isa<>`

vabridgers commented:
Hi @steakhal , this change seems to have exposed by div/0 error in a very particular corner we came across after integrating this change. Could you try this case to see if can repro?

The div/0 is occurring at line 1093: 
       bool IncompleteLastElement = (NumBytesRead % ElemSizeInChars) != 0;

clang -cc1 -analyze -analyzer-checker=unix.Stream -analyzer-opt-analyze-headers test.c 

This is the test source. 

# 1 "" 3
typedef FILE;
void *b;
a() {
  FILE f = fopen("", "");
  fread(b, 1, 1, f);
}

The crash ... 
`$ clang -cc1 -analyze -analyzer-checker=unix.Stream -analyzer-opt-analyze-headers test.c`
`PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.`
`Stack dump:`
`0.      Program arguments: clang -cc1 -analyze -analyzer-checker=unix.Stream -analyzer-opt-analyze-headers test.c`
`1.      <eof> parser at end of file`
`2.      While analyzing stack:`
`        #0 Calling a`
`3.      :5:3: Error evaluating statement`
`4.      :5:3: Error evaluating statement`
` #0 0x000000000686046a llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) llvm/lib/Support/Unix/Signals.inc:723:22`
` #1 0x00000000068608a4 PrintStackTraceSignalHandler(void*) llvm/lib/Support/Unix/Signals.inc:798:1`
` #2 0x000000000685e1bd llvm::sys::RunSignalHandlers() llvm/lib/Support/Signals.cpp:105:20`
` #3 0x000000000685fe54 SignalHandler(int) llvm/lib/Support/Unix/Signals.inc:413:1`
` #4 0x00007f931d2c1630 __restore_rt sigaction.c:0:0`
` #5 0x0000000009fc96ff tryToInvalidateFReadBufferByElements(llvm::IntrusiveRefCntPtr<clang::ento::ProgramState const>, clang::ento::CheckerContext&, clang::ento::CallEvent const&, clang::ento::NonLoc, clang::ento::NonLoc) clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp:1093:48`

steakhal commented:
> Hi @steakhal , this change seems to have exposed by div/0 error in a very particular corner we came across after integrating this change. Could you try this case to see if can repro?

Yea, it seems to crash. I'll fix it once I have some time. Probably early next week if I can't spare some time during the weekend.
Thanks for raising this.
